packages/datadog-instrumentations/src/grpc/client.js

```diff
@@ -101,71 +103,71 @@ function wrapMethod (method, path, type) {
   return wrapped
 }
 
-function wrapCallback (requestResource, parentResource, callback) {
+function wrapCallback (ctx, callback = () => { }) {
   return function (err) {
     if (err) {
-      requestResource.runInAsyncScope(() => {
-        errorChannel.publish(err)
-      })
+      ctx.error = err
+      errorChannel.publish(ctx)
     }
 
-    if (callback) {
-      return parentResource.runInAsyncScope(() => {
-        return callback.apply(this, arguments)
-      })
-    }
+    return asyncStartChannel.runStores(ctx, () => {
+      return callback.apply(this, arguments)
+      // No async end channel needed
+    })
   }
 }
 
-function wrapStream (call, requestResource, parentResource) {
-  if (!call || typeof call.emit !== 'function') return
-
-  shimmer.wrap(call, 'emit', emit => {
-    return function (eventName, ...args) {
-      requestResource.runInAsyncScope(() => {
-        switch (eventName) {
-          case 'error':
-            errorChannel.publish(args[0])
-
-            break
-          case 'status':
-            finishChannel.publish(args[0])
-
-            break
-        }
-      })
+function createWrapEmit (ctx) {
+  return function wrapEmit (emit) {
+    return function (event, arg1) {
+      switch (event) {
```

Comment:
These might run in the wrong context with the current approach, and should be handled in a `bindStore` instead.

Reply:
It would be simpler to just alter the error handler to pull from the ctx object rather than using `this.activeSpan`. The `finish` function already does that anyway. No point wasting perf on manipulating store context when we control all the code it's running. 🤷🏻

Reply:
I opted for just making it use the explicit span stored on the `ctx`. No need to manipulate the store that way so better performance and more reliable.

Reply:
This could work for the client, but it doesn't work for servers, since which context to use depends on the event.

Reply:
That doesn't seem to be accurate, unless the existing instrumentation was incorrect. All events ran in the `requestStore` context, only the callback used the `parentStore`. This is the same behaviour as what I have here.

Reply:
I don't think this is correct, but at the same time, fixing it is probably out of scope of this PR. I still think a normalized approach for event emitters would be easier to follow, and easier to add as a helper to TracingChannel.

Reply:
As discussed offline, let's move this PR forward to fix the immediate issue and revisit event emitters soon after.